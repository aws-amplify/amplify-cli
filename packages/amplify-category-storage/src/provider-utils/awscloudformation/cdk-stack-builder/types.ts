import { AmplifyCDKL1 } from '@aws-amplify/cli-extensibility-helper';
import * as cdk from '@aws-cdk/core';
import { $TSObject } from 'amplify-cli-core';
import { DdbAttrType } from '../cfn-template-utils';

export interface AmplifyDDBResourceInputParameters {
  tableName: string;
  partitionKeyName: string;
  partitionKeyType: DdbAttrType;
  sortKeyName?: string;
  sortKeyType?: DdbAttrType;
}

export type AmplifyCfnParamType = {
  params: Array<string>;
  paramType: string;
  default?: string;
};

export interface AmplifyS3ResourceInputParameters {
  bucketName?: string;
  resourceName?: string;
  policyUUID?: string;
  authPolicyName?: string;
  unauthPolicyName?: string;
  authRoleName?: $TSObject;
  unauthRoleName?: $TSObject;
  s3PublicPolicy?: string;
  s3PrivatePolicy?: string; //default:"NONE"
  s3ProtectedPolicy?: string;
  s3UploadsPolicy?: string;
  s3ReadPolicy?: string;
  s3PermissionsAuthenticatedPublic?: string;
  s3PermissionsAuthenticatedProtected?: string;
  s3PermissionsAuthenticatedPrivate?: string;
  s3PermissionsAuthenticatedUploads?: string;
  s3PermissionsGuestPublic?: string;
  s3PermissionsGuestUploads?: string;
  AuthenticatedAllowList?: string;
  GuestAllowList?: string;
  selectedGuestPermissions?: Array<string>;
  selectedAuthenticatedPermissions?: Array<string>;
  triggerFunction?: string;
  adminTriggerFunction?: string;
}

//Base class for all storage resource stacks ( S3, DDB )
export class AmplifyResourceCfnStack extends cdk.Stack implements AmplifyCDKL1 {
  _cfnParameterMap: Map<string, cdk.CfnParameter> = new Map();
  constructor(scope: cdk.Construct, id: string) {
    super(scope, id, undefined);
  }

  /**
   *
   * @param props :cdk.CfnOutputProps
   * @param logicalId: : lodicalId of the Resource
   */
  addCfnOutput(props: cdk.CfnOutputProps, logicalId: string): void {
    try {
      new cdk.CfnOutput(this, logicalId, props);
    } catch (error) {
      throw error;
    }
  }

  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnMapping(props: cdk.CfnMappingProps, logicalId: string): void {
    try {
      new cdk.CfnMapping(this, logicalId, props);
    } catch (error) {
      throw error;
    }
  }

  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnCondition(props: cdk.CfnConditionProps, logicalId: string): void {
    try {
      new cdk.CfnCondition(this, logicalId, props);
    } catch (error) {
      throw error;
    }
  }
  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnResource(props: cdk.CfnResourceProps, logicalId: string): cdk.CfnResource {
    try {
      return new cdk.CfnResource(this, logicalId, props);
    } catch (error) {
      throw error;
    }
  }
  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnParameter(props: cdk.CfnParameterProps, logicalId: string): void {
    try {
      if (this._cfnParameterMap.has(logicalId)) {
        throw new Error('logical Id already Exists');
      }
      this._cfnParameterMap.set(logicalId, new cdk.CfnParameter(this, logicalId, props));
    } catch (error) {
      throw error;
    }
  }

  //Generate convert cdk stack to cloudformation
  public renderCloudFormationTemplate = (): string => {
    return this._toCloudFormation();
  };
}

//Types used in Build/Params.json
export enum AmplifyBuildParamsPermissions {
  ALLOW = 'ALLOW',
  DISALLOW = 'DISALLOW',
}
