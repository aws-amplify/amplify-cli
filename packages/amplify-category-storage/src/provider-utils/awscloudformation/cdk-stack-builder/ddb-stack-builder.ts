import * as cdk from '@aws-cdk/core';
import * as ddb from '@aws-cdk/aws-dynamodb';
import { DynamoDBCLIInputs, DynamoDBCLIInputsKeyType, FieldType } from '../service-walkthrough-types/dynamoDB-user-input-types';
import { AmplifyDDBResourceTemplate } from '@aws-amplify/cli-extensibility-helper';

const CFN_TEMPLATE_FORMAT_VERSION = '2010-09-09';
const ROOT_CFN_DESCRIPTION = 'DDB Resource for AWS Amplify CLI';

export class AmplifyDDBResourceStack extends cdk.Stack implements AmplifyDDBResourceTemplate {
  _scope: cdk.Construct;
  dynamoDBTable!: ddb.CfnTable;
  _props: DynamoDBCLIInputs;
  _cfnParameterMap: Map<string, cdk.CfnParameter> = new Map();

  constructor(scope: cdk.Construct, id: string, props: DynamoDBCLIInputs) {
    super(scope, id, undefined);
    this._scope = scope;
    this._props = props;
    this.templateOptions.templateFormatVersion = CFN_TEMPLATE_FORMAT_VERSION;
    this.templateOptions.description = ROOT_CFN_DESCRIPTION;
  }

  /**
   *
   * @param props :cdk.CfnOutputProps
   * @param logicalId: : lodicalId of the Resource
   */
  addCfnOutput(props: cdk.CfnOutputProps, logicalId: string): void {
    try {
      new cdk.CfnOutput(this, logicalId, props);
    } catch (error) {
      throw new Error(error);
    }
  }

  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnMapping(props: cdk.CfnMappingProps, logicalId: string): void {
    try {
      new cdk.CfnMapping(this, logicalId, props);
    } catch (error) {
      throw new Error(error);
    }
  }

  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnCondition(props: cdk.CfnConditionProps, logicalId: string): void {
    try {
      new cdk.CfnCondition(this, logicalId, props);
    } catch (error) {
      throw new Error(error);
    }
  }
  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnResource(props: cdk.CfnResourceProps, logicalId: string): void {
    try {
      new cdk.CfnResource(this, logicalId, props);
    } catch (error) {
      throw new Error(error);
    }
  }

  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnParameter(props: cdk.CfnParameterProps, logicalId: string): void {
    try {
      if (this._cfnParameterMap.has(logicalId)) {
        throw new Error('logical Id already Exists');
      }
      this._cfnParameterMap.set(logicalId, new cdk.CfnParameter(this, logicalId, props));
    } catch (error) {
      throw new Error(error);
    }
  }

  generateStackResources = async () => {
    let usedAttributes: DynamoDBCLIInputsKeyType[] = [];
    let keySchema: ddb.CfnTable.KeySchemaProperty[] = [];
    let globalSecondaryIndexes: ddb.CfnTable.GlobalSecondaryIndexProperty[] = [];

    if (this._props.partitionKey) {
      usedAttributes.push(this._props.partitionKey);
      keySchema.push({
        attributeName: this._props.partitionKey.fieldName,
        keyType: 'HASH',
      });
    }
    if (this._props.sortKey) {
      usedAttributes.push(this._props.sortKey);
      keySchema.push({
        attributeName: this._props.sortKey.fieldName,
        keyType: 'RANGE',
      });
    }
    if (this._props.gsi && this._props.gsi.length > 0) {
      this._props.gsi.forEach(gsi => {
        let gsiIndex = {
          indexName: gsi.name,
          keySchema: [
            {
              attributeName: gsi.partitionKey.fieldName,
              keyType: 'HASH',
            },
          ],
          projection: {
            projectionType: 'ALL',
          },
          provisionedThroughput: {
            readCapacityUnits: 5,
            writeCapacityUnits: 5,
          },
        };

        if (usedAttributes.findIndex(attr => attr.fieldName === gsi.partitionKey.fieldName) === -1) {
          usedAttributes.push(gsi.partitionKey);
        }
        if (gsi.sortKey) {
          gsiIndex.keySchema.push({
            attributeName: gsi.sortKey?.fieldName,
            keyType: 'RANGE',
          });
          if (usedAttributes.findIndex(attr => attr?.fieldName === gsi.sortKey?.fieldName) == -1) {
            usedAttributes.push(gsi.sortKey);
          }
        }
        globalSecondaryIndexes.push(gsiIndex);
      });
    }

    const ddbAttrTypeMapping = {
      string: 'S',
      number: 'N',
      binary: 'B',
      boolean: 'BOOL',
      list: 'L',
      map: 'M',
      null: 'NULL',
      'string-set': 'SS',
      'number-set': 'NS',
      'binary-set': 'BS',
    };

    let attributeMapping: ddb.CfnTable.AttributeDefinitionProperty[] = [];

    usedAttributes.forEach((attr: DynamoDBCLIInputsKeyType) => {
      attributeMapping.push({
        attributeName: attr.fieldName,
        attributeType: ddbAttrTypeMapping[attr.fieldType],
      });
    });

    this.dynamoDBTable = new ddb.CfnTable(this, 'DynamoDBTable', {
      tableName: cdk.Fn.conditionIf(
        'ShouldNotCreateEnvResources',
        cdk.Fn.ref('tableName'),
        cdk.Fn.join('', [cdk.Fn.ref('tableName'), '-', cdk.Fn.ref('env')]),
      ).toString(),
      attributeDefinitions: attributeMapping,
      keySchema,
      globalSecondaryIndexes,
      provisionedThroughput: {
        readCapacityUnits: 5,
        writeCapacityUnits: 5,
      },
      streamSpecification: {
        streamViewType: 'NEW_IMAGE',
      },
    });
  };

  public renderCloudFormationTemplate = (): string => {
    return JSON.stringify(this._toCloudFormation(), undefined, 2);
  };
}
