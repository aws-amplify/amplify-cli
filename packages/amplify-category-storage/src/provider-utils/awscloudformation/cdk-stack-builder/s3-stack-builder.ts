// import * as cdk from '@aws-cdk/core';
// import * as s3 from '@aws-cdk/aws-s3';
// import * as iam from '@aws-cdk/aws-iam';
// import { AmplifyRootStackTemplate } from '';
// import { IStackSynthesizer, ISynthesisSession } from '@aws-cdk/core';
