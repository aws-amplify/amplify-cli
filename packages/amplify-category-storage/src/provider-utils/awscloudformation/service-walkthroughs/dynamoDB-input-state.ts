import {
  $TSAny,
  $TSObject,
  AmplifyCategories,
  AmplifySupportedService,
  CLIInputSchemaValidator,
  JSONUtilities,
  pathManager,
} from 'amplify-cli-core';
import * as fs from 'fs-extra';
import * as path from 'path';
import { getFieldType } from '../cfn-template-utils';
import { DynamoDBCLIInputs, DynamoDBCLIInputsGSIType } from '../service-walkthrough-types/dynamoDB-user-input-types';

/* Need to move this logic to a base class */

export class DynamoDBInputState {
  _cliInputsFilePath: string; //cli-inputs.json (output) filepath
  _resourceName: string; //user friendly name provided by user
  _category: string; //category of the resource
  _service: string; //AWS service for the resource
  buildFilePath: string;

  constructor(resourceName: string) {
    this._category = AmplifyCategories.STORAGE;
    this._service = AmplifySupportedService.DYNAMODB;
    this._resourceName = resourceName;

    const projectBackendDirPath = pathManager.getBackendDirPath();
    this._cliInputsFilePath = path.resolve(path.join(projectBackendDirPath, AmplifyCategories.STORAGE, resourceName, 'cli-inputs.json'));
    this.buildFilePath = path.resolve(path.join(projectBackendDirPath, AmplifyCategories.STORAGE, resourceName, 'build'));
  }

  public getCliInputPayload(): DynamoDBCLIInputs {
    let cliInputs: DynamoDBCLIInputs;

    // Read cliInputs file if exists
    try {
      cliInputs = JSONUtilities.readJson<DynamoDBCLIInputs>(this._cliInputsFilePath)!;
    } catch (e) {
      throw new Error('cli-inputs.json file missing from the resource directory');
    }

    return cliInputs;
  }

  public cliInputFileExists(): boolean {
    return fs.existsSync(this._cliInputsFilePath);
  }

  public isCLIInputsValid(cliInputs?: DynamoDBCLIInputs) {
    if (!cliInputs) {
      cliInputs = this.getCliInputPayload();
    }

    const schemaValidator = new CLIInputSchemaValidator(this._service, this._category, 'DynamoDBCLIInputs');
    schemaValidator.validateInput(JSON.stringify(cliInputs));
  }

  public saveCliInputPayload(cliInputs: DynamoDBCLIInputs): void {
    this.isCLIInputsValid(cliInputs);

    fs.ensureDirSync(pathManager.getResourceDirectoryPath(undefined, this._category, this._resourceName));
    try {
      JSONUtilities.writeJson(this._cliInputsFilePath, cliInputs);
    } catch (e) {
      throw new Error(e);
    }
  }

  public migrate() {
    let cliInputs: DynamoDBCLIInputs;

    // migrate the resource to new directory structure if cli-inputs.json is not found for the resource

    const backendDir = pathManager.getBackendDirPath();
    const oldParametersFilepath = path.join(backendDir, 'storage', this._resourceName, 'parameters.json');
    const oldCFNFilepath = path.join(backendDir, 'storage', this._resourceName, `${this._resourceName}-cloudformation-template.json`);
    const oldStorageParamsFilepath = path.join(backendDir, 'storage', this._resourceName, `storage-params.json`);

    const oldParameters = JSONUtilities.readJson<$TSAny>(oldParametersFilepath, { throwIfNotExist: true });
    const oldCFN = JSONUtilities.readJson<$TSAny>(oldCFNFilepath, { throwIfNotExist: true });
    const oldStorageParams = JSONUtilities.readJson<$TSAny>(oldStorageParamsFilepath, { throwIfNotExist: false }) || {};

    const partitionKey = {
      fieldName: oldParameters.partitionKeyName,
      fieldType: getFieldType(oldParameters.partitionKeyType),
    };

    let sortKey;

    if (oldParameters.sortKeyName) {
      sortKey = {
        fieldName: oldParameters.sortKeyName,
        fieldType: getFieldType(oldParameters.sortKeyType),
      };
    }

    let triggerFunctions = [];

    if (oldStorageParams.triggerFunctions) {
      triggerFunctions = oldStorageParams.triggerFunctions;
    }

    const getType = (attrList: $TSAny, attrName: string) => {
      let attrType;

      attrList.forEach((attr: $TSAny) => {
        if (attr.AttributeName === attrName) {
          attrType = getFieldType(attr.AttributeType);
        }
      });

      return attrType;
    };

    let gsi: DynamoDBCLIInputsGSIType[] = [];

    if (oldCFN?.Resources?.DynamoDBTable?.Properties?.GlobalSecondaryIndexes) {
      oldCFN.Resources.DynamoDBTable.Properties.GlobalSecondaryIndexes.forEach((cfnGSIValue: $TSAny) => {
        let gsiValue: $TSAny = {};
        (gsiValue.name = cfnGSIValue.IndexName),
          cfnGSIValue.KeySchema.forEach((keySchema: $TSObject) => {
            if (keySchema.KeyType === 'HASH') {
              gsiValue.partitionKey = {
                fieldName: keySchema.AttributeName,
                fieldType: getType(oldCFN.Resources.DynamoDBTable.Properties.AttributeDefinitions, keySchema.AttributeName),
              };
            } else {
              gsiValue.sortKey = {
                fieldName: keySchema.AttributeName,
                fieldType: getType(oldCFN.Resources.DynamoDBTable.Properties.AttributeDefinitions, keySchema.AttributeName),
              };
            }
          });
        gsi.push(gsiValue);
      });
    }
    cliInputs = {
      resourceName: this._resourceName,
      tableName: oldParameters.tableName,
      partitionKey,
      sortKey,
      triggerFunctions,
      gsi,
    };

    this.saveCliInputPayload(cliInputs);

    // Remove old files

    if (fs.existsSync(oldCFNFilepath)) {
      fs.removeSync(oldCFNFilepath);
    }
    if (fs.existsSync(oldParametersFilepath)) {
      fs.removeSync(oldParametersFilepath);
    }
    if (fs.existsSync(oldStorageParamsFilepath)) {
      fs.removeSync(oldStorageParamsFilepath);
    }
  }
}
