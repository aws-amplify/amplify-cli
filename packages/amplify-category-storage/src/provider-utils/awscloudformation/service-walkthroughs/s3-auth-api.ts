import { $TSAny, $TSContext } from 'amplify-cli-core';
import { printer } from 'amplify-prompts';
import { AmplifyCategories } from 'amplify-cli-core';
import os from 'os';

/* This file contains all functions interacting with AUTH category */

//UPSTREAM API: function to be called from Storage to fetch or update Auth resources

/**
 * Get the name of the Auth resource used by S3
 * @param context  used to fetch all auth resources used by storage(S3)
 * @returns Name of the auth resource used by S3
 */
export async function getAuthResourceARN(context: $TSContext): Promise<string> {
  let authResources = (await context.amplify.getResourceStatus('auth')).allResources;
  authResources = authResources.filter((resource: $TSAny) => resource.service === 'Cognito');
  if (authResources.length === 0) {
    throw new Error('No auth resource found. Please add it using amplify add auth');
  }
  return authResources[0].resourceName as string;
}
/**
 * Migrate all Auth resources used by Storage(S3) for Override feature.
 * @param context - used to fetch auth resources and to migrate auth resources for override-feature.
 * @return migration successful?
 */
export async function migrateAuthDependencyResource(context: $TSContext): Promise<boolean> {
  let authResourceName = undefined;
  try {
    authResourceName = await getAuthResourceARN(context);
  } catch (error) {
    // No auth resources to migrate - new project
    return true;
  }
  if (authResourceName) {
    return context.amplify.invokePluginMethod<boolean>(context, AmplifyCategories.AUTH, undefined, 'migrateAuthResource', [
      context,
      authResourceName,
    ]);
  }
  return true;
}

/**
 * Check if storage authentication requirements are satisfied by the configured storage
 * @param context
 * @param storageResourceName
 * @param allowUnauthenticatedIdentities
 */
export async function checkStorageAuthenticationRequirements(
  context: $TSContext,
  storageResourceName: string,
  allowUnauthenticatedIdentities: boolean,
) {
  const storageRequirements = { authSelections: 'identityPoolAndUserPool', allowUnauthenticatedIdentities };

  const checkResult: $TSAny = await context.amplify.invokePluginMethod(context, AmplifyCategories.AUTH, undefined, 'checkRequirements', [
    storageRequirements,
    context,
    'storage',
    storageResourceName,
  ]);

  // If auth is imported and configured, we have to throw the error instead of printing since there is no way to adjust the auth
  // configuration.
  if (checkResult.authImported === true && checkResult.errors && checkResult.errors.length > 0) {
    throw new Error(checkResult.errors.join(os.EOL));
  }

  if (checkResult.errors && checkResult.errors.length > 0) {
    printer.warn(checkResult.errors.join(os.EOL));
  }

  // If auth is not imported and there were errors, adjust or enable auth configuration
  if (!checkResult.authEnabled || !checkResult.requirementsMet) {
    try {
      // If this is not set as requirement, then explicitly configure it to disabled.
      if (storageRequirements.allowUnauthenticatedIdentities === undefined) {
        storageRequirements.allowUnauthenticatedIdentities = false;
      }

      await context.amplify.invokePluginMethod(context, AmplifyCategories.AUTH, undefined, 'externalAuthEnable', [
        context,
        AmplifyCategories.STORAGE,
        storageResourceName,
        storageRequirements,
      ]);
    } catch (error) {
      printer.error(error as string);
      throw error;
    }
  }
}
