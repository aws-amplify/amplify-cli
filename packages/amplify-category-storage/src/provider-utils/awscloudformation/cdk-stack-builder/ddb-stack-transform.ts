import { AmplifyDDBResourceTemplate } from '@aws-amplify/cli-extensibility-helper';
import * as cdk from '@aws-cdk/core';
import { App } from '@aws-cdk/core';
import { $TSAny, buildOverrideDir, JSONUtilities, pathManager } from 'amplify-cli-core';
import { formatter, printer } from 'amplify-prompts';
import * as fs from 'fs-extra';
import os from 'os';
import * as path from 'path';
import * as vm from 'vm2';
import { DynamoDBCLIInputs } from '../service-walkthrough-types/dynamoDB-user-input-types';
import { DynamoDBInputState } from '../service-walkthroughs/dynamoDB-input-state';
import { AmplifyDDBResourceStack } from './ddb-stack-builder';
import { AmplifyDDBResourceInputParameters } from './types';
import { getDdbAttrType } from '../cfn-template-utils';

export class DDBStackTransform {
  app: App;
  _cliInputs: DynamoDBCLIInputs;
  _resourceTemplateObj: AmplifyDDBResourceStack | undefined;
  _cliInputsState: DynamoDBInputState;
  _cfn!: string;
  _cfnInputParams!: AmplifyDDBResourceInputParameters;
  _resourceName: string;

  constructor(resourceName: string) {
    this.app = new App();
    this._resourceName = resourceName;

    // Validate the cli-inputs.json for the resource
    this._cliInputsState = new DynamoDBInputState(resourceName);
    this._cliInputs = this._cliInputsState.getCliInputPayload();
    this._cliInputsState.isCLIInputsValid();
  }

  async transform() {
    // Generate  cloudformation stack from cli-inputs.json
    await this.generateStack();

    // Generate  cloudformation stack input params from cli-inputs.json
    this.generateCfnInputParameters();

    // Modify cloudformation files based on overrides
    await this.applyOverrides();

    // Save generated cloudformation.json and parameters.json files
    this.saveBuildFiles();
  }

  generateCfnInputParameters() {
    this._cfnInputParams = {
      tableName: this._cliInputs.tableName,
      partitionKeyName: this._cliInputs.partitionKey.fieldName,
      partitionKeyType: getDdbAttrType(this._cliInputs.partitionKey.fieldType),
    };
    if (this._cliInputs.sortKey) {
      this._cfnInputParams.sortKeyName = this._cliInputs.sortKey.fieldName;
      this._cfnInputParams.sortKeyType = getDdbAttrType(this._cliInputs.sortKey.fieldType);
    }
  }

  async generateStack() {
    this._resourceTemplateObj = new AmplifyDDBResourceStack(this.app, 'AmplifyDDBResourceStack', this._cliInputs);

    // Add Parameters
    this._resourceTemplateObj.addCfnParameter(
      {
        type: 'String',
      },
      'partitionKeyName',
    );
    this._resourceTemplateObj.addCfnParameter(
      {
        type: 'String',
      },
      'partitionKeyType',
    );
    this._resourceTemplateObj.addCfnParameter(
      {
        type: 'String',
      },
      'env',
    );
    if (this._cliInputs.sortKey) {
      this._resourceTemplateObj.addCfnParameter(
        {
          type: 'String',
        },
        'sortKeyName',
      );

      this._resourceTemplateObj.addCfnParameter(
        {
          type: 'String',
        },
        'sortKeyType',
      );
    }
    this._resourceTemplateObj.addCfnParameter(
      {
        type: 'String',
      },
      'tableName',
    );

    // Add conditions

    this._resourceTemplateObj.addCfnCondition(
      {
        expression: cdk.Fn.conditionEquals(cdk.Fn.ref('env'), 'NONE'),
      },
      'ShouldNotCreateEnvResources',
    );

    // Add resources

    await this._resourceTemplateObj.generateStackResources();

    // Add outputs
    this._resourceTemplateObj.addCfnOutput(
      {
        value: cdk.Fn.ref('DynamoDBTable'),
      },
      'Name',
    );
    this._resourceTemplateObj.addCfnOutput(
      {
        value: cdk.Fn.getAtt('DynamoDBTable', 'Arn').toString(),
      },
      'Arn',
    );
    this._resourceTemplateObj.addCfnOutput(
      {
        value: cdk.Fn.getAtt('DynamoDBTable', 'StreamArn').toString(),
      },
      'StreamArn',
    );
    this._resourceTemplateObj.addCfnOutput(
      {
        value: cdk.Fn.ref('partitionKeyName'),
      },
      'PartitionKeyName',
    );
    this._resourceTemplateObj.addCfnOutput(
      {
        value: cdk.Fn.ref('partitionKeyType'),
      },
      'PartitionKeyType',
    );

    if (this._cliInputs.sortKey) {
      this._resourceTemplateObj.addCfnOutput(
        {
          value: cdk.Fn.ref('sortKeyName'),
        },
        'SortKeyName',
      );
      this._resourceTemplateObj.addCfnOutput(
        {
          value: cdk.Fn.ref('sortKeyType'),
        },
        'SortKeyType',
      );
    }

    this._resourceTemplateObj.addCfnOutput(
      {
        value: cdk.Fn.ref('AWS::Region'),
      },
      'Region',
    );
  }

  async applyOverrides() {
    const backendDir = pathManager.getBackendDirPath();
    const resourceDirPath = pathManager.getResourceDirectoryPath(undefined, 'storage', this._resourceName);
    const overrideJSFilePath = path.resolve(path.join(resourceDirPath, 'build', 'override.js'));

    const isBuild = await buildOverrideDir(backendDir, resourceDirPath).catch(error => {
      printer.error(`Build error : ${error.message}`);
      throw new Error(error);
    });
    // skip if packageManager or override.ts not found
    if (isBuild) {
      const { override } = await import(overrideJSFilePath).catch(error => {
        formatter.list(['No override File Found', `To override ${this._resourceName} run amplify override auth ${this._resourceName} `]);
        return undefined;
      });

      if (typeof override === 'function' && override) {
        const overrideCode: string = await fs.readFile(overrideJSFilePath, 'utf-8').catch(() => {
          formatter.list(['No override File Found', `To override ${this._resourceName} run amplify override auth`]);
          return '';
        });

        const sandboxNode = new vm.NodeVM({
          console: 'inherit',
          timeout: 5000,
          sandbox: {},
          require: {
            context: 'sandbox',
            builtin: ['path'],
            external: true,
          },
        });
        try {
          await sandboxNode.run(overrideCode, overrideJSFilePath).override(this._resourceTemplateObj as AmplifyDDBResourceTemplate);
        } catch (err: $TSAny) {
          const error = new Error(`Skipping override due to ${err}${os.EOL}`);
          printer.error(`${error}`);
          error.stack = undefined;
          throw error;
        }
      }
    }
  }

  saveBuildFiles() {
    if (this._resourceTemplateObj) {
      this._cfn = JSON.parse(this._resourceTemplateObj.renderCloudFormationTemplate());
    }

    // store files in local-filesysten

    fs.ensureDirSync(this._cliInputsState.buildFilePath);
    const cfnFilePath = path.resolve(path.join(this._cliInputsState.buildFilePath, `${this._resourceName}-cloudformation-template.json`));
    try {
      JSONUtilities.writeJson(cfnFilePath, this._cfn);
    } catch (e) {
      throw new Error(e);
    }

    fs.ensureDirSync(this._cliInputsState.buildFilePath);
    const cfnInputParamsFilePath = path.resolve(path.join(this._cliInputsState.buildFilePath, 'parameters.json'));
    try {
      JSONUtilities.writeJson(cfnInputParamsFilePath, this._cfnInputParams);
    } catch (e) {
      throw new Error(e);
    }
  }
}
