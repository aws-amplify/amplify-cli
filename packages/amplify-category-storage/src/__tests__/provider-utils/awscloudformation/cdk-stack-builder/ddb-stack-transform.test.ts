/* These tests test the DDBStackTransform and run the cdk builder tool which is used within this file */

import { DDBStackTransform } from '../../../../provider-utils/awscloudformation/cdk-stack-builder/ddb-stack-transform';
import {
  DynamoDBCLIInputs,
  FieldType,
} from '../../../../provider-utils/awscloudformation/service-walkthrough-types/dynamoDB-user-input-types';
import { DynamoDBInputState } from '../../../../provider-utils/awscloudformation/service-walkthroughs/dynamoDB-input-state';

jest.mock('amplify-cli-core', () => ({
  buildOverrideDir: jest.fn().mockResolvedValue(false),
  JSONUtilities: {
    writeJson: jest.fn(),
    readJson: jest.fn(),
  },
  pathManager: {
    getBackendDirPath: jest.fn().mockReturnValue('mockbackendpath'),
    getResourceDirectoryPath: jest.fn().mockReturnValue('mockresourcepath'),
  },
}));
jest.mock('fs-extra', () => ({
  readFileSync: () => jest.fn().mockReturnValue('{ "Cognito": { "provider": "aws"}}'),
  existsSync: () => jest.fn().mockReturnValue(true),
  ensureDirSync: jest.fn().mockReturnValue(true),
}));

jest.mock('path', () => ({
  join: jest.fn().mockReturnValue('mockjoinedpath'),
  resolve: jest.fn().mockReturnValue('mockjoinedpath'),
}));

jest.mock('../../../../provider-utils/awscloudformation/service-walkthroughs/dynamoDB-input-state');

describe('Test DDB transform generates correct CFN template', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Generated ddb template with all CLI configurations set with no overrides', async () => {
    const resourceName = 'mockResource';
    const cliInputsJSON: DynamoDBCLIInputs = {
      resourceName: resourceName,
      tableName: 'mocktablename',
      partitionKey: {
        fieldName: 'id',
        fieldType: FieldType.string,
      },
      sortKey: {
        fieldName: 'name',
        fieldType: FieldType.number,
      },
      gsi: [
        {
          name: 'gsiname',
          partitionKey: {
            fieldName: 'name',
            fieldType: FieldType.number,
          },
        },
        {
          name: 'updategsiname',
          partitionKey: {
            fieldName: 'col',
            fieldType: FieldType.string,
          },
        },
      ],
      triggerFunctions: [],
    };

    jest.spyOn(DynamoDBInputState.prototype, 'getCliInputPayload').mockImplementation(() => cliInputsJSON);
    const ddbTransform = new DDBStackTransform(resourceName);
    await ddbTransform.transform();

    console.log(ddbTransform._cfn);
    console.log(ddbTransform._cfnInputParams);

    expect(ddbTransform._cfn).toMatchSnapshot();
    expect(ddbTransform._cfnInputParams).toMatchSnapshot();
  });
});
