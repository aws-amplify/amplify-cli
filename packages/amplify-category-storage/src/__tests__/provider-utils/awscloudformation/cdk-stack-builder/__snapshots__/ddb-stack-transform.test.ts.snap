// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test DDB transform generates correct CFN template Generated ddb template with all CLI configurations set with no overrides 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Conditions": Object {
    "ShouldNotCreateEnvResources": Object {
      "Fn::Equals": Array [
        Object {
          "Ref": "env",
        },
        "NONE",
      ],
    },
  },
  "Description": "DDB Resource for AWS Amplify CLI",
  "Outputs": Object {
    "Arn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "DynamoDBTable",
          "Arn",
        ],
      },
    },
    "Name": Object {
      "Value": Object {
        "Ref": "DynamoDBTable",
      },
    },
    "PartitionKeyName": Object {
      "Value": Object {
        "Ref": "partitionKeyName",
      },
    },
    "PartitionKeyType": Object {
      "Value": Object {
        "Ref": "partitionKeyType",
      },
    },
    "Region": Object {
      "Value": Object {
        "Ref": "AWS::Region",
      },
    },
    "SortKeyName": Object {
      "Value": Object {
        "Ref": "sortKeyName",
      },
    },
    "SortKeyType": Object {
      "Value": Object {
        "Ref": "sortKeyType",
      },
    },
    "StreamArn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "DynamoDBTable",
          "StreamArn",
        ],
      },
    },
  },
  "Parameters": Object {
    "env": Object {
      "Type": "String",
    },
    "partitionKeyName": Object {
      "Type": "String",
    },
    "partitionKeyType": Object {
      "Type": "String",
    },
    "sortKeyName": Object {
      "Type": "String",
    },
    "sortKeyType": Object {
      "Type": "String",
    },
    "tableName": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "DynamoDBTable": Object {
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "name",
            "AttributeType": "N",
          },
          Object {
            "AttributeName": "col",
            "AttributeType": "S",
          },
        ],
        "GlobalSecondaryIndexes": Array [
          Object {
            "IndexName": "gsiname",
            "KeySchema": Array [
              Object {
                "AttributeName": "name",
                "KeyType": "HASH",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
            "ProvisionedThroughput": Object {
              "ReadCapacityUnits": 5,
              "WriteCapacityUnits": 5,
            },
          },
          Object {
            "IndexName": "updategsiname",
            "KeySchema": Array [
              Object {
                "AttributeName": "col",
                "KeyType": "HASH",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
            "ProvisionedThroughput": Object {
              "ReadCapacityUnits": 5,
              "WriteCapacityUnits": 5,
            },
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "name",
            "KeyType": "RANGE",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "StreamSpecification": Object {
          "StreamViewType": "NEW_IMAGE",
        },
        "TableName": Object {
          "Fn::If": Array [
            "ShouldNotCreateEnvResources",
            Object {
              "Ref": "tableName",
            },
            Object {
              "Fn::Join": Array [
                "",
                Array [
                  Object {
                    "Ref": "tableName",
                  },
                  "-",
                  Object {
                    "Ref": "env",
                  },
                ],
              ],
            },
          ],
        },
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
}
`;

exports[`Test DDB transform generates correct CFN template Generated ddb template with all CLI configurations set with no overrides 2`] = `
Object {
  "partitionKeyName": "id",
  "partitionKeyType": "S",
  "sortKeyName": "name",
  "sortKeyType": "N",
  "tableName": "mocktablename",
}
`;
